### PR TITLE
UHF-3615: School groups

### PR DESCRIPTION
**Pull request by tapsal**
# Groups for general upper secondary schools [UHF-3615](https://helsinkisolutionoffice.atlassian.net/browse/UHF-3615)
Makes it possible to add groups for general upper secondary schools. With a new school editor role, it's possible to limit which users can edit the content within the groups.

## What was done
* Enabled group and group_content_menu contrib modules.
* Added a group type for general upper secondary schools.
* Added new school editor role.
* Added support for group TPR units with link action to help adding existing units.
* Added views page for managing TPR units within the group.
* The group menu uses the same layout as the normal sidebar menu.

## How to install
* Make sure your instance is up and running.
* Use this branch.
    * `git pull origin UHF-4163-school-groups`
* Get template changes for the theme:
    * `composer require drupal/hdbt:dev-UHF-4163-group-menus` 
* Get admin theme changes:
    * `composer require drupal/hdbt_admin:dev-UHF-4163-group-menus` 
* Run `make shell` and then:
    * `drush cim`
    * `drush cr`

## How to test
Create a new school group using admin user (a user with an admin role).
* Go to `/admin/group` -> Add group.
  * Add group name, e.g. "Ressun lukio -ryhmä".
* Create a front page for the new school group.
  * From the group page -> Nodes -> Add new content.
  * Use the school name as the name of the node.
  * Give the node a meaningful path alias (e.g. /lukiot/ressun-lukio).
  * Add the node to the group menu.
* Add an existing TPR school unit (e.g. Ressun lukio) to the group.
  * From the group page -> Units -> Add TPR unit.
  * Edit the unit page:
    * Override title to "Yhteystiedot".
    * Add menu link to group menu. Use previously created school front page as parent.
    * Publish the unit.
* Add a user with "School editor" role and add the new user to the school group.
  * From the group page -> Add member.
  * You can give the user a "School admin" group role so that the user is able to add and remove members.

Now login with the school editor user and test that you can edit and add new nodes to the group. Also add the the new nodes to the group's menu.

## Other PRs
* https://github.com/City-of-Helsinki/drupal-hdbt/pull/258
* https://github.com/City-of-Helsinki/drupal-hdbt-admin/pull/113

### CODE REVIEW

File: conf/cmi/group.role.upper_secondary_school-0b0baeca9.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+uuid: 0326015a-c353-48b6-8544-e45d8925cdd2
+langcode: fi
```
Comment (teroelonen) [inline conf/cmi/group.role.upper_secondary_school-0b0baeca9.yml:2]: This should be "en".

File: conf/cmi/group.role.upper_secondary_school-97a902313.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+langcode: en
+status: true
+dependencies:
+  config:
+    - group.type.upper_secondary_school
+  enforced:
+    config:
+      - user.role.read_only
+id: upper_secondary_school-97a902313
+label: 'Read only'
```
Comment (teroelonen) [inline conf/cmi/group.role.upper_secondary_school-97a902313.yml:11]: I wonder if these roles will be broken now that the dev-branch has the role changes (new roles & changes for names)

File: conf/cmi/user.role.content_producer.yml [hunk 3]
```diff
@@ -14,6 +14,7 @@ dependencies:
   module:
     - content_translation
     - file
+    - group
```
Comment (teroelonen) [inline conf/cmi/user.role.content_producer.yml:17]: These need to be merged with the new roles as well.

File: public/modules/custom/helfi_group/src/Plugin/GroupContentEnabler/GroupTprUnit.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ *   entity_access = TRUE,
+ *   reference_label = @Translation("Title"),
+ *   reference_description = @Translation("The title of the TPR Unit to add to the group"),
+ *   handlers = {
+ *     "access" = "Drupal\group\Plugin\GroupContentAccessControlHandler",
+ *     "permission_provider" = "Drupal\group\Plugin\GroupContentPermissionProvider",
+ *   }
+ * )
+ */
+class GroupTprUnit extends GroupContentEnablerBase {
```
Comment (teroelonen) [inline public/modules/custom/helfi_group/src/Plugin/GroupContentEnabler/GroupTprUnit.php:26]: Why is this empty? I might just ask dumb questions here 😄

File: conf/cmi/core.extension.yml [hunk 5]
```diff
@@ -36,6 +36,9 @@ module:
   flysystem_azure: 0
   focal_point: 0
   gin_toolbar: 0
+  gnode: 0
```
Comment (teroelonen) [inline conf/cmi/core.extension.yml:39]: What is this?

File: conf/cmi/core.extension.yml [hunk 6]
```diff
@@ -114,6 +118,7 @@ module:
   update: 0
   update_helper: 0
   user: 0
+  variationcache: 0
```
Comment (teroelonen) [inline conf/cmi/core.extension.yml:121]: What is this?

**Review by teroelonen — CHANGES_REQUESTED** [review]
I left some questions for the code :)

File: public/modules/custom/helfi_group/src/Plugin/GroupContentEnabler/GroupTprUnit.php (hunk 4 above)
Comment (tapsal) [reply public/modules/custom/helfi_group/src/Plugin/GroupContentEnabler/GroupTprUnit.php:26]: The functionality is defined with the GroupContentEnabler annotation.

File: conf/cmi/core.extension.yml (hunk 5 above)
Comment (tapsal) [reply conf/cmi/core.extension.yml:39]: It's the Group Node module, which enables group functionality for nodes.

File: conf/cmi/core.extension.yml (hunk 6 above)
Comment (tapsal) [reply conf/cmi/core.extension.yml:121]: It's a requirement for the Group module. https://www.drupal.org/project/variationcache

File: conf/cmi/group.role.upper_secondary_school-0b0baeca9.yml (hunk 1 above)
Comment (tapsal) [reply conf/cmi/group.role.upper_secondary_school-0b0baeca9.yml]: Fixed.

File: conf/cmi/group.role.upper_secondary_school-97a902313.yml (hunk 2 above)
Comment (tapsal) [reply conf/cmi/group.role.upper_secondary_school-97a902313.yml:11]: Checking and adding permissions regarding new roles.

File: conf/cmi/user.role.content_producer.yml (hunk 3 above)
Comment (tapsal) [reply conf/cmi/user.role.content_producer.yml:22]: Checking and adding permissions regarding new roles.

**Review by teroelonen — APPROVED** [review]
Tested and worked as promised 🦖

**Pull request merged**